### ISSUE TITLE
Mark self as mut borrow in solver methods?

### ISSUE THREAD

**Issue by 108anup** [prove-rs/z3.rs#266]
I am new to rust, but I noticed that the functions like `Solver::assert` borrow an immutable reference to `self`. Should these be annotated as mutable given that such functions modify the solver object. If not, could you please explain the rationale behind it.

```
pub fn assert(&self, ast: &ast::Bool<'ctx>) {
    ///       ^^^^^ Here
    debug!("assert: {:?}", ast);
    unsafe { Z3_solver_assert(self.ctx.z3_ctx, self.z3_slv, ast.z3_ast) };
}
```

**Comment by Pat-Lafon** [comment]
I think this is best described as the [Interior Mutability Pattern](https://doc.rust-lang.org/book/ch15-05-interior-mutability.html#refcellt-and-the-interior-mutability-pattern) since all objects are secretly reference counted on the Z3 c side of things.

**Comment by 108anup** [comment]
Okay, I guess the reference counting can ensure memory leaks are absent. But that may not protect against race conditions, right?

For example, one might borrow a solver for its model in a closure. Then one might update the solver assertions, and then query the closure. If now the closure is called, then the result is for the updated solver.

The onus is then on the user to ensure that such things don't happen. There are no compile-time or run-time checks for this, right? If the annotation is made mutable, then such behaviors would not be allowed by the compiler. What is the benefit of keeping the annotation immutable?

In the above example, the closure could just borrow the model and not the solver to avoid any inconsistency. But ideally IMO, the compiler should force the user to do this and this should not be a choice because isn't such safety the whole point of rust, otherwise one might as well use C++.

**Comment by waywardmonkeys** [comment]
IIRC, the underlying C++ code is already thread safe which is why we were able to remove the `Z3_MUTEX` that existed in earlier versions of the code.

If you start to use `&mut self`, you also make it harder to use some of this from multiple threads (which I think some people do or did do...)

**Comment by Pat-Lafon** [comment]
How does one write this program? I'm not super familiar with threads so more for my own benefit. So I think almost all of the major structures in the z3 crate are `!Send` and `!Sync` because raw pointers are one of the exceptions for auto-deriving these(https://doc.rust-lang.org/nomicon/send-and-sync.html). Thus you can't just trivially pass a reference into a spawned thread without a compiler error.

You can isolate everything into it's own thread, but that seems safe to me anyways given z3's guarantees here.
```rust
use std::{
    thread::{self, sleep},
    time::Duration,
};

use z3::{
    ast::{Ast, Int},
    Config, Context, Solver,
};

fn main() {
    let mut handles = Vec::new();
    for _ in 0..5 {
        let handle = thread::spawn(move || {
            let cfg = &Config::new();
            let ctx = Context::new(cfg);
            let x = Int::new_const(&ctx, "x");
            let solver = Solver::new(&ctx);
            sleep(Duration::from_secs(1));
            solver.assert(&x._eq(&x));
            sleep(Duration::from_secs(1));
            solver.check();
        });
        handles.push(handle);
    }

    for t in handles.into_iter() {
        t.join().unwrap();
    }

    println!("Hello, world!");
}
```

**Comment by toolCHAINZ** [comment]
I find myself somewhat persuaded by the argument here. As stated, not a memory safety issue, but it _would_ give developers a way to signal how the solver ought to be used in their own code, which right now is not possible.

I don't think this change would impact "most"  usage of the library? I looked at my own usage of the library and I'm only ever asserting things in cases where I own the solver, or where I'm easily able to give a mut ref. My usage has solvers mostly short-lived whereas Asts and Models stick around; I'm not sure how "standard" this is per others' usage.

I'm also not sure of the use-case waywardmonkeys was referring to regarding threading as I'm pretty sure it was explicitly not supported before #404

**Comment by Pat-Lafon** [comment]
It's a little bit less clear to me what the win is here. I'm a little worried this would then make the library inconsistent, if one doesn't also have context used under a `&mut` as well?

I'll note that I've sometimes (but not saying it's pervasive) immediately passed solvers/contexts only by reference, not by ownership. This is merely to avoid putting `&` everywhere because the solver/context is already a reference. This still falls into your usage pattern so it wouldn't be a huge change to fix...

**Comment by Pat-Lafon** [comment]
Would you limit this to assert, or would check/push/pop/set_params take the solver by `&mut` as well?

On the other hand, I see that add_assign is currently implemented with a `&mut solver` which I didn't know about(fun fact?).

**Comment by toolCHAINZ** [comment]
I think if we were to do this, I think it would make sense to do on all of those other than check. 

For motivation , I could see a case where a user might want to expose a solver in a callback, say for logging or something, but assure that no additional assertions are made to it. A strained use case to be sure, but I could maybe see it happening.

I guess the add assign is that way because the trait forces it?

But yeah, I don't have a concrete use case from my own work that would actually benefit from this, so maybe as you say, it's better to not do this.

**Comment by Pat-Lafon** [comment]
> For motivation , I could see a case where a user might want to expose a solver in a callback, say for logging or something, but assure that no additional assertions are made to it. A strained use case to be sure, but I could maybe see it happening.

I can't fully envision this but could a user impose this constraint on themselves without the bindings needing to do it? Since if you pass the solver by ownership or `&mut` into the callback, you can still use it as `&`.

> I guess the add assign is that way because the trait forces it?

Yeah, it's required by the trait, showing that the interface isn't as nicely consistent currently as I had thought.

**Comment by toolCHAINZ** [comment]
Yeah, I think we'll say we're not doing this for now at least. If a motivating real-world use case turns up showing how this would actually be beneficial, we can re-evaluate.
